### PR TITLE
Makes Destroy() call parent & misc cleanup

### PR DESCRIPTION
**Pull request by ike709**
## About The Pull Request

Adds `SHOULD_CALL_PARENT(TRUE)` to `Destroy()`. Frankly I hadn't realized we were missing this until kube pointed it on in Discord.

The three really important procs (weakrefs, global vars, and qdel() itself) have changes ripped straight from latest TG. The rest is pretty much just me slapping the relevant parent calls in and doing misc cleanup.

## Changelog
:cl:
refactor: Destroy() now always calls parent.
refactor: Cleaned up a fair bit of SSAugury, but more work is needed.
refactor: Miscellaneous minor code improvements & optimizations.
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/fire_burning.dm [hunk 1]
```diff
@@ -27,7 +27,7 @@ SUBSYSTEM_DEF(fire_burning)
 	while(currentrun.len)
 		var/obj/O = currentrun[currentrun.len]
 		currentrun.len--
-		if (!O || QDELETED(O))
+		if (QDELETED(O))
```
Comment (ike709) [inline code/controllers/subsystem/fire_burning.dm:30]: The first check in the macro is literally `!O`

File: code/controllers/subsystem/job.dm [hunk 2]
```diff
@@ -106,7 +106,7 @@ SUBSYSTEM_DEF(job)
 		var/datum/job/job = GetJob(rank)
 		if(!job)
 			return FALSE
-		if(is_banned_from(player.ckey, rank) || QDELETED(player))
+		if(QDELETED(player) || is_banned_from(player.ckey, rank))
```
Comment (ike709) [inline code/controllers/subsystem/job.dm:109]: Potentially short circuit on a fairly quick check instead of running an expensive proc first.

File: code/datums/position_point_vector.dm [hunk 3]
```diff
@@ -207,7 +207,8 @@
 	var/paused = FALSE
 
 /datum/point/vector/processed/Destroy()
-	STOP_PROCESSING(SSprojectiles, src)
+	..()
+	return PROCESS_KILL
```
Comment (ike709) [inline code/datums/position_point_vector.dm]:
```suggestion
	STOP_PROCESSING(SSprojectiles, src)
	return ..()
```

Comment (ike709) [inline code/datums/position_point_vector.dm]:
```suggestion
	STOP_PROCESSING(SSprojectiles, src)
	return ..()
```

File: code/modules/shuttle/super_cruise/orbital_poi_generator/loot/alien_artifact.dm [hunk 4]
```diff
@@ -89,8 +89,9 @@
 		START_PROCESSING(SSobj, src)
 
 /datum/artifact_effect/Destroy()
+	. = ..()
```
Comment (ike709) [inline code/modules/shuttle/super_cruise/orbital_poi_generator/loot/alien_artifact.dm]:
```suggestion
```

File: code/modules/shuttle/super_cruise/orbital_poi_generator/loot/alien_artifact.dm [hunk 5]
```diff
@@ -89,8 +89,9 @@
...
 	if(requires_processing)
-		STOP_PROCESSING(SSobj, src)
+		return PROCESS_KILL
```
Comment (ike709) [inline code/modules/shuttle/super_cruise/orbital_poi_generator/loot/alien_artifact.dm]:
```suggestion
		STOP_PROCESSING(SSobj, src)

	return ..()
```

File: code/controllers/subsystem/augury.dm [hunk 6]
```diff
@@ -18,40 +18,37 @@ SUBSYSTEM_DEF(augury)
...
-		for(var/i in GLOB.player_list)
-			if(isobserver(i) && (!(observers_given_action[i])))
+	if(length(doombringers))
+		for(var/mob/dead/observer/O in GLOB.player_list)
+			if(!(O in observers_given_action))
 				var/datum/action/innate/augury/A = new
-				A.Grant(i)
-				observers_given_action[i] = TRUE
+				A.Grant(O)
+				observers_given_action[O]
```
Comment (ike709) [inline code/controllers/subsystem/augury.dm]:
```suggestion
				observers_given_action += O
```

File: code/controllers/subsystem/mapping.dm [hunk 7]
```diff
@@ -157,12 +157,11 @@ SUBSYSTEM_DEF(mapping)
...
 	for(var/datum/d in nuke_threats)
-		if(!istype(d) || QDELETED(d))
+		if(QDELETED(d))
 			nuke_threats -= d
 
-	for(var/N in nuke_tiles)
-		var/turf/open/floor/circuit/C = N
-		C.update_icon()
+	for(var/turf/open/floor/circuit/C as() in nuke_tiles)
+		C?.update_icon()
```
Comment (ike709) [inline code/controllers/subsystem/mapping.dm]:
```suggestion
		C.update_icon()
```

File: code/controllers/subsystem/augury.dm [hunk 8]
```diff
@@ -18,40 +18,37 @@ SUBSYSTEM_DEF(augury)
...
 			continue
 		var/threat = doombringers[d]
 		if((biggest_threat == null) || (biggest_threat < threat))
 			biggest_doom = d
 			biggest_threat = threat
 
-	if(doombringers.len)
-		for(var/i in GLOB.player_list)
-			if(isobserver(i) && (!(observers_given_action[i])))
+	if(length(doombringers))
```
Comment (Archanial) [inline code/controllers/subsystem/augury.dm:33]: Isn't looping indexes more performant than looping observers then checking if they're in the list?

Comment (ike709) [reply code/controllers/subsystem/augury.dm:33]: Considering I'm nuking this entire subsystem as soon as I get my open PR backlog cleared, I wouldn't worry about it.

File: code/controllers/subsystem/garbage.dm [hunk 9]
```diff
@@ -284,12 +284,6 @@ SUBSYSTEM_DEF(garbage)
 // Should be treated as a replacement for the 'del' keyword.
 // Datums passed to this will be given a chance to clean up references to allow the GC to collect them.
 /proc/qdel(datum/D, force=FALSE, ...)
-	if(isweakref(D))
-		var/datum/weakref/weakref = D
-		D = weakref.resolve()
-		if(!D)
-			return
```
Comment (Archanial) [inline code/controllers/subsystem/garbage.dm:291]: I really hope this doesn't break anything.
What's the reason for this change?

Comment (ike709) [reply code/controllers/subsystem/garbage.dm:291]: Literally just got moved to `/datum/weakref/Destroy(force)` so it stops typechecking every datum

**Review by Archanial — APPROVED** [review]
